### PR TITLE
Add samples for azure active directory b2c work with spring security

### PR DESCRIPTION
**Pull request by zhichengliu12581**
There are some functions in Azure Active Directory B2C that cannot be directly implemented by spring security (B2C can assign `appRole` to the application, specify the page to clear the cache after B2C logout, etc.), these functions need to add some special configurations. It may take some time for users to find these configurations, so we hope to add these samples so that users can more clearly know how to use Azure Active Directory B2C with spring security.

This pr only provides the simplest sample, these functions already have similar samples. However, we expect to provide these samples to describe how spring security and b2c work together. Perhaps these samples can better show users how B2C and spring security are work together.

### CODE REVIEW

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc [hunk 1]
```diff
@@ -0,0 +1,74 @@
+= Azure Active Directory B2C with Spring Security Sample
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc:1]: Tutorial of using Azure Active Directory B2C by Spring Security

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+In the current folder, we have the these samples to introduce how to use the Azure Active Directory B2C functions through Spring Security.
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc:3]: By reading this tutorial, you will learn how to use Azure Active Directory B2C by Spring Security.

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc (hunk 1 above)
Comment (chenrujun) [reply servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc:1]: Add toc in line 2:
```
:toc:
```
Refs: https://docs.asciidoctor.org/asciidoc/latest/toc/

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+The following sections provide the samples for Azure Active Directory B2C work with Spring Security.
+
+* <<work-as-web-application, Work as Web Application>>
+* <<work-as-resource-server, Work as Resource Server>>
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc:8]: We can delete this after we have `:toc:`

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+[[create-azure-active-directory-B2C-resources]]
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc:10]: Before create resource, we should tell customer how to get a Azure account. I mean we need a `prerequisites`

Example:

https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/spring/ENVIRONMENT_CHECKLIST.md#prerequisites

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+Follow the guide of https://docs.microsoft.com/en-us/azure/active-directory-b2c/add-profile-editing-policy?pivots=b2c-user-flow[AAD B2C set up a profile edit flow].
+
+=== Enable password reset
+
+Follow the guide of https://docs.microsoft.com/en-us/azure/active-directory-b2c/add-password-reset-policy?pivots=b2c-user-flow#create-a-password-reset-user-flow[AAD B2C set up a password reset user flow]
+
+[[work-as-web-application]]
+== Work as web application
+
+After prepare Azure Active Directory B2C resources, we can use this <<../azure-active-directory-b2c/web-application-1-simple-sample/, sample>> to use it as a web application.
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc:42]: Sample project: <<./web-application-1-simple-sample/, web-application-1-simple-sample>>

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+=== Setting the redirect URI
+
+In order to jump to the correct page (home) after `profile-edit` and `password-reset`, we need to add `http://localhost:8080/` in the redirect uri.
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc:46]: This already exist in `Register your Azure Active Directory B2C application`

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+With the guide https://docs.microsoft.com/en-us/azure/active-directory-b2c/tutorial-register-applications?tabs=app-reg-ga#register-a-web-application[AAD-B2C-application-registry] and step 7.
+
+=== Application config
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc:49]: Fill necessary contents in sample project.

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+You are then redirected to the default _auto-generated_ login page. You can sign in or sign up with your account.
+
+Then, you can edit your profile with the button `Profile edit` or reset your password with `Password reset`.
+
+[[work-as-resource-server]]
+== Work as Resource Server
+
+After prepare Azure Active Directory B2C resources, we can use this <<../azure-active-directory-b2c/resource-server-1-simple-sample/, sample>> to use it as a web application.
+
+=== Application config
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc:67]: Fill necessary contents in sample project.

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+1. Fill necessary contents in `application.yml`, like `${your-jwk-set-uri}`.
+
+2. Fill necessary contents in `home.html`, like `${your-tenant-name}`.
+
+=== Boot up the application
+
+Launch the Spring Boot 2.0 sample and go to `http://localhost:8080`.
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc:57]: Lunch the sample project, then open `http://localhost:8080` by browser.

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+== Work as Resource Server
+
+After prepare Azure Active Directory B2C resources, we can use this <<../azure-active-directory-b2c/resource-server-1-simple-sample/, sample>> to use it as a web application.
+
+=== Application config
+
+1. Fill necessary contents in `application.yml`, like `${your-jwk-set-uri}`.
+
+=== Boot up the application
+
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/README.adoc:72]: Lunch the sample project. Then open `http://localhost:8080/getResourceServer` by browser. The browser will access `web-application-1-simple-sample`, and `web-application-1-simple-sample` will access `resource-server-1-simple-server` by access token.

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/resource-server-1-simple-sample/README.adoc [hunk 11]
```diff
@@ -0,0 +1,3 @@
+= Azure Active Directory B2C with Spring Security Sample for resource server
+
+Please refer to this <<../README.adoc#work-as-resource-server,document>>
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/resource-server-1-simple-sample/README.adoc:3]: nit: `document` -> `tutorial`.

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/resource-server-1-simple-sample/src/main/java/com/spring/sample/b2c/controller/SampleController.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+@Controller
+public class SampleController {
+
+    @ResponseBody
+    @GetMapping(value = { "/hello" })
+    public String hello() {
+        return "Hello, this is aad-b2c-resource-server-simple-sample.";
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/resource-server-1-simple-sample/src/main/java/com/spring/sample/b2c/controller/SampleController.java:16]: `this is a resource-server protected by Azure Active Directory B2C. ` Then we can reuse this if we adding other resource server projects.

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/web-application-1-simple-sample/README.adoc [hunk 13]
```diff
@@ -0,0 +1,3 @@
+= Azure Active Directory B2C with Spring Security Sample for resource server
+
+Please refer to this <<../README.adoc#work-as-web-application,document>>
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/web-application-1-simple-sample/README.adoc:3]: Same here.

File: servlet/spring-boot/java/oauth2/azure-active-directory-b2c/web-application-1-simple-sample/src/main/java/com/spring/sample/b2c/controller/SampleController.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Autowired
+    private WebClient webClient;
+
+    @GetMapping(value = { "/", "/home" })
+    public String index() {
+        return "home";
+    }
+
+    @GetMapping(value = { "/getResourceServer" })
```
Comment (chenrujun) [inline servlet/spring-boot/java/oauth2/azure-active-directory-b2c/web-application-1-simple-sample/src/main/java/com/spring/sample/b2c/controller/SampleController.java:24]: `/getResourceServer` -> `/resourceServer`

Not use verb in url.

Refs: https://cloud.google.com/blog/products/api-management/restful-api-design-nouns-are-good-verbs-are-bad

**Review by chenrujun — CHANGES_REQUESTED** [review]

**Comment by jzheaux** [comment]
> This pr only provides the simplest sample, these functions already have similar samples.

This was my first thought as well when looking over the PR. I wonder if it would be better to first contribute to the READMEs.

For example, you could add to [the OAuth 2.0 Login setup instructions](https://github.com/spring-projects/spring-security-samples/tree/main/servlet/spring-boot/java/oauth2/login#login-with-okta), [the WebClient instructions](https://github.com/spring-projects/spring-security-samples/tree/main/servlet/spring-boot/java/oauth2/webclient#oauth-20-webclient-servlet-sample), etc.

**Comment by zhichengliu12581** [comment]
@jzheaux Thank you very much for your review.
I added the configuration description of B2C in [OAuth 2.0 Login Setting Instructions](https://github.com/spring-projects/spring-security-samples/tree/main/servlet/spring-boot/java/oauth2/login).

As a resource server, I intend to use it as the simplest example of use. Because other examples will be added later, including the use of custom `JwtDecoder`, `DelegatingJwtGrantedAuthoritiesConverter`, etc., and how to configure it in B2C. It may take some time for users to configure these things in B2C, and some progressive samples may be convenient for users to understand how to configure.

The same is true as a client. Some functions (such as adding role or configuring logout) will be more convenient for users to configure if there is a sample to show.

If you agree, I will add more examples of how to use B2C. New features will be gradually added between them.

**Comment by zhichengliu12581** [comment]
@jzheaux Can you please help to review this pr again.
I added some samples, and there are some small differences between them.
Maybe by adding chapters, and each chapter contains a small change compare to previous chapter can make user easily understand the usage of each feature.
I may need to modify some descriptions, but please see if it is appropriate to add such samples.

**Comment by zhichengliu12581** [comment]
@jzheaux ping

**Comment by jzheaux** [comment]
>  It may take some time for users to configure these things in B2C, and some progressive samples may be convenient for users to understand how to configure.

While I agree with this, I don't think `spring-security-samples` is the right home for it. You can imagine if each vendor wanted to add their own chapters of samples into `spring-security-samples` for any given module and see that it would quickly become unmaintainable.

Additionally, since Azure uses their own set of properties, I think this could end up being confusing to new users perusing this repository.

For now, I'd leave the PR at just adding README instructions that get things working with the Spring Boot default properties. If you have your own repo that points to additional samples, I think it would be reasonable to include a link to it in the README.

**Comment by marcusdacoregio** [comment]
I'm closing this since there weren't any updates since @jzheaux last comment. We can always reopen it if the suggested changes are addressed.

Thanks!

**Pull request closed without merging**